### PR TITLE
NFI cleanup and support blocking calls

### PR DESCRIPTION
**Pull request by eregon**

### CODE REVIEW

**Comment by eregon** [comment]
This is blocked until SVM tests use a proper and built TruffleRuby home.

File: src/main/ruby/core/rubinius.rb [hunk 1]
```diff
@@ -44,24 +44,29 @@ module Rubinius
 
   HOST_OS = Truffle::System.host_os
 
+  IS_BSD = HOST_OS == 'bsd'
+  IS_LINUX = HOST_OS == 'linux'
+  IS_SOLARIS = HOST_OS == 'solaris'
```
Comment (nirvdrum) [inline src/main/ruby/core/rubinius.rb]: These checks aren't the same as the ones they're replacing. Did you verify they work the same on various BSDs, Solaris, and Linux distributions? If not, I'd switch back to the regexp match.

Comment (eregon) [reply src/main/ruby/core/rubinius.rb]: This is all already normalized by `Truffle::System.host_os`. If not, it should be fixed there.

Comment (eregon) [reply src/main/ruby/core/rubinius.rb]: These work for the Solaris, Darwin and Linux we have in CI at the very least.

Comment (eregon) [reply src/main/ruby/core/rubinius.rb]: Maybe we can check that the value of `host_os` is within one of the expected values we know about, in `Truffle::System.host_os`.

Comment (nirvdrum) [reply src/main/ruby/core/rubinius.rb]: I see. Looking at `BasicPlatform` it looks like the `IS_BSD` value would be wrong, but we don't officially support any BSD either.

File: src/main/java/org/truffleruby/core/thread/ThreadManager.java [hunk 2]
```diff
@@ -408,6 +408,20 @@ public int runBlockingSystemCallUntilResult(Node currentNode, BlockingAction<Int
...
     }
 
+    @TruffleBoundary
+    public int runBlockingNFISystemCallUntilResult(Node currentNode, BlockingAction<Integer> action) {
+        assert Errno.EINTR.defined();
+        int EINTR = Errno.EINTR.intValue();
+
+        return runUntilResult(currentNode, () -> {
+            int result = action.block();
+            if (result == -EINTR) {
```
Comment (bjfish) [inline src/main/java/org/truffleruby/core/thread/ThreadManager.java]: What is `-EINTR` ?

Comment (eregon) [reply src/main/java/org/truffleruby/core/thread/ThreadManager.java]: A special value to indicate the call failed with errno set to EINTR.

Comment (bjfish) [reply src/main/java/org/truffleruby/core/thread/ThreadManager.java]: Okay, maybe this would be more clear as `EINTR_FAIL = -EINTR`

Comment (eregon) [reply src/main/java/org/truffleruby/core/thread/ThreadManager.java]: The only caller is Ruby, and they need to agree on a convention. I'm not sure extracting this to a constant and duplicating the constant is better.

File: src/main/ruby/core/rubinius.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/core/rubinius.rb]: I fixed this by checking the 3 possibles platforms that match BSD, and removing the "unknown" platform entirely, have a look at 69e441c.

File: src/main/java/org/truffleruby/core/thread/ThreadManager.java (hunk 2 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/thread/ThreadManager.java]: Done, I called it `INTERRUPTED`.

**Pull request merged**